File: 3Sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {

        List<List<Integer>> random = new ArrayList<>() ;

        if (nums.length < 3) return random ;

        Set<List<Integer>> s = new HashSet<>() ;

        Integer n = nums.length ;
        Arrays.sort(nums) ;


        for ( Integer i = 0 ; i < nums.length - 2 ; i++ ) {
            Integer j = i + 1 ;
            Integer k = n - 1 ;
            while (j < k) {
                List<Integer> dummy = new ArrayList<>() ;
                Integer sum = nums[i] + nums[j] + nums[k] ;
                if (sum == 0) {
                    dummy.add(nums[i]) ;
                    dummy.add(nums[j]) ;
                    dummy.add(nums[k]) ;
                    j++ ; k-- ;
                    s.add(dummy) ;
                    while (j < k && nums[j] == nums[j - 1]) j++ ;
                    while (j < k && nums[k] == nums[k + 1]) k-- ;
                } else if (sum < 0) {
                    j++ ;
                } else if (sum > 0) {
                    k-- ;
                }
            }
        }

        List<List<Integer>> result = new ArrayList<>(s);
        return result ;
    }
}

File: 3Sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

        vector<vector<int>> random ;

        if (nums.size() < 3) return random ;

        set<vector<int>> s ;

        int n = nums.size() ;

        sort(nums.begin(), nums.end()) ;

        for ( int i = 0 ; i < n - 2 ; i++ ) {
            int j = i + 1 ;
            int k = n - 1 ;
            while (j < k) {
                int sum = nums[i] + nums[j] + nums[k] ;
                if (sum == 0) {
                    vector<int> dummy ;
                    dummy.push_back(nums[i]) ;
                    dummy.push_back(nums[j]) ;
                    dummy.push_back(nums[k]) ;
                    s.insert(dummy);
                    j++;
                    k--;
                    while (j < k && nums[j] == nums[j - 1]) j++ ;
                    while (j < k && nums[k] == nums[k + 1]) k-- ;
                }
                else if (sum < 0) {
                    j++ ;
                }
                else if (sum > 0) {
                    k-- ;
                }
            }
        }

        vector<vector<int>> res(s.begin(), s.end()) ;
        return res ;
    }
};